### PR TITLE
core: add call_preinitcalls()

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Adds call_preinitcalls() for really early initcalls. This function is
supposed to be called before call_initcalls() is called. With
virtualization enabled it is called in a blocking context when the
OP-TEE partition is created.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

This PR breaks out the initcall stuff from https://github.com/OP-TEE/optee_os/pull/4844

I hope to continue the discussion here. I believe that the hooks enabled with with are needed, but the names may have become a bit messy with the extensions added over the time. Perhaps we can find some better name.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Here's some parts from the previous discussion:

> @jenswi-linaro : I'm taking the comments on `core/include/initcall.h` out of context here since I suppose this can take a wihle to sort out.
> 
> > @etienne-lms : `bootfinal` callbacks are called before `*_init` and `preinit` levels. It sounds strange to me
> 
> If we disregard the preinit levels this is an accurate picture of the master branch status.
> 
> > @etienne-lms : Looking at the registered `boot_final` callbacks, i think they are misused:
> > ```
> > core/arch/arm/kernel/secure_partition.c:boot_final(sp_init_all);
> > core/arch/arm/kernel/boot.c:boot_final(release_external_dt);
> > core/drivers/crypto/se050/glue/i2c.c:boot_final(load_trampoline);
> > ```
> > The `secure_partition.c` callbacks could be `early_init` level, there are no dependencies.
> 
> This depends on what will happen with SPs and `CFG_VIRTUALIZATION=y`. I don't think anyone have given much thought on that. I suspect in case it will ever be enabled with virtualization it will either be managed from the nexus or a special partition.
> 
> >  @etienne-lms : The `se050/i2c.c` callback should be called only once a VM has been created, no? (`driver_init_late` level)
> 
> `boot_final()` was added to handle the I2C case, but I'm afraid that the virtualization case was overlooked. What will happen with I2C with virtualization? My guess is that one of the partitions will claim ownership. The same partition that matches the guest in normal world which takes ownership of the physical I2C controller. So for now I'd say that `load_trampoline` isn't compatible with virtualization.
> 
> > @etienne-lms : As for `release_external_dt()`, maybe it should be simply explicitly called from `boot_init_primary_late()` instead of `call_finalcalls()`.
> 
> `boot_final()` was agreed to be the last thing done before exiting to normal world for the first time, this a very good match for what's needed for `release_external_dt` regardless of virtualization.

File: core/include/initcall.h [hunk 1]
```diff
@@ -30,12 +30,67 @@ struct initcall {
...
+ *  | 1. call_preinitcalls()        | In the nexus                      |
+ *  | 2. call_initcalls()           +-----------------------------------+
+ *  | 3. call_finalcalls()          | 1. call_finalcalls()              |
+ *  +-------------------------------+-----------------------------------+
+ *  | "Primary CPU switching to normal world boot" is printed           |
+ *  +-------------------------------+-----------------------------------+
+ *                                  | A guest is created and            |
+ *                                  | virt_guest_created() is called.   |
+ *                                  | After the partition has been      |
+ *                                  | created and activiated.           |
```
Comment (jforissier) [inline core/include/initcall.h:78]: activated

File: core/include/initcall.h [hunk 2]
```diff
@@ -30,12 +30,67 @@ struct initcall {
...
+ *                                  +-----------------------------------+
+ *                                  | 2. call_preinitcalls()            |
+ *                                  +-----------------------------------+
+ *                                  | When the partition is recieving   |
```
Comment (jforissier) [inline core/include/initcall.h:82]: receiving

**Comment by jforissier** [comment]
> I like the initcall way of registering functions to be called at a certain place, it's decentralized and has a minimal overhead.

Agreed.

> The biggest problem may be to give the different places good names to avoid confusion.

Yes, not an easy task especially since good names cannot just reflect the order in which the functions are called, due to the differences between virt and non-virt. The big comment block in this PR is very helpful in that respect.

> Should we try to sort out this now or wait until all the other pieces needed by https://linaro.atlassian.net/browse/STR-23 are in place?

IMO any suggestions are welcome at any time, we can try to think about better names now, but the current names are not really a problem and even less so if your description is merged. Also, the names are purely internal so no compatibility issue etc.

With the typos fixed this LGTM:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

Note, I prefer to merge things when they are actually used, but I also understand it is desirable to split big PRs like 4844. Could we at least have some of the refactoring that uses the preinit calls?

**Comment by jenswi-linaro** [comment]
Update

**Pull request merged**